File: 33. Search in Rotated Sorted Array.cpp
/*
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

You are given a target value to search. If found in the array return its index, otherwise return -1.

You may assume no duplicate exists in the array.

Subscribe to see which companies asked this question
*/
class Solution {
public:
	int search(vector<int>& nums, int target) {
		if (nums.size() != 0)
		{
			unsigned int first = 0;
			unsigned int last = nums.size() - 1;
			do
			{
				unsigned int mid = (first + last) / 2;
				if (nums[mid] == target) 
					return mid;
					
				if (nums[first] < nums[mid])
					if (nums[first] <= target && target <= nums[mid])
						last = mid - 1;
					else
						first = mid + 1;
				else
					if (nums[mid] < target && target <= nums[last])
						first = mid + 1;
					else
						last = mid - 1;
			} while (first != last + 1);
			return -1;
		}
		else
			return -1;
	}
};

File: Sorting and Searching/33. Search in Rotated Sorted Array.py
# Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

# (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

# You are given a target value to search. If found in the array return its index, otherwise return -1.

# You may assume no duplicate exists in the array.

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) / 2
            if nums[mid] == target:
                return mid
            else:
                if nums[mid] < nums[right]:
                    # right side is ascending
                    if nums[mid] < target <= nums[right]:
                        left = mid + 1
                    else:
                        right = mid - 1
                else:
                    # left side is ascending
                    if nums[left] <= target < nums[mid]:
                        right = mid - 1
                    else:
                        left = mid + 1
        return -1
